packages/react-storage/src/components/StorageBrowser/views/LocationsView/useLocationsView.ts

```diff
@@ -0,0 +1,116 @@
+import React from 'react';
+
+import { useLocationsData } from '../../do-not-import-from-here/actions';
+import { usePaginate } from '../hooks/usePaginate';
+import { LocationData } from '../../actions';
+import { useStore } from '../../providers/store';
+
+interface UseLocationsView {
+  hasNextPage: boolean;
+  hasError: boolean;
+  isLoading: boolean;
+  isPaginateNextDisabled: boolean;
+  isPaginatePreviousDisabled: boolean;
+  message: string | undefined;
+  pageItems: LocationData[];
+  page: number;
+  onNavigate: (location: LocationData) => void;
+  onRefresh: () => void;
+  onPaginateNext: () => void;
+  onPaginatePrevious: () => void;
+}
+
+interface InitialValues {
+  pageSize?: number;
+}
+
+export type LocationsViewActionType =
+  | { type: 'REFRESH_DATA' }
+  | { type: 'RESET' }
+  | { type: 'PAGINATE_NEXT' }
+  | { type: 'PAGINATE_PREVIOUS' }
+  | { type: 'PAGINATE'; page: number }
+  | { type: 'NAVIGATE'; location: LocationData }
+  | { type: 'SEARCH'; query: string };
+
+export interface UseLocationsViewOptions {
+  initialValues?: InitialValues;
+  onDispatch?: React.Dispatch<LocationsViewActionType>;
+  onNavigate?: (location: LocationData) => void;
+}
+
+const DEFAULT_PAGE_SIZE = 100;
+export const DEFAULT_LIST_OPTIONS = {
+  exclude: 'WRITE' as const,
+  pageSize: DEFAULT_PAGE_SIZE,
+};
+
+export function useLocationsView(
+  options?: UseLocationsViewOptions
+): UseLocationsView {
+  const [state, handleList] = useLocationsData();
+  const [, dispatchStoreAction] = useStore();
+  const { data, message, hasError, isLoading } = state;
+  const { result, nextToken } = data;
+  const resultCount = result.length;
+  const hasNextToken = !!nextToken;
+
+  const onNavigate = options?.onNavigate;
+  const initialValues = options?.initialValues ?? {};
+
+  const [listOptions] = React.useState({
+    ...DEFAULT_LIST_OPTIONS,
+    ...initialValues,
+  });
+
+  // initial load
+  React.useEffect(() => {
+    handleList({
+      options: { ...listOptions, refresh: true },
+    });
+  }, [handleList, listOptions]);
```

Comment:
`defaultPageSize` can be put inside a ref since there will be no changes to it that require re-rendering, e.g.
```ts
  const defaultListOptions = React.useRef({ ...DEFAULT_LIST_OPTIONS, pageSize: defaultPageSize });

  React.useEffect(() => {
    handleList({
      options: { ...defaultListOptions.current, refresh: true },
    });
  }, [handleList]);
```
Also if there is no use case for `DEFAULT_LIST_OPTIONS` besides this call we can move `refresh` inside of it as well